cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/ETLSparkProgram.java

```diff
@@ -140,32 +131,104 @@ public Object call(Tuple2<String, Object> input) throws Exception {
+
+    Preconditions.checkArgument(sparkTransformName.equals(Iterables.getOnlyElement(sourceNextStages)),
+                                "Expected the single stage after the source stage to be the spark transform: %s",
+                                sparkTransformName);
+
+
+    // phase starting from source to spark transform
+    PipelinePhase sourcePhase = phaseSpec.getPhase().subsetTo(ImmutableSet.of(sparkTransformName));
+    String sourcePipelineStr =
+      GSON.toJson(new BatchPhaseSpec(phaseSpec.getPhaseName(), sourcePhase, phaseSpec.getResources(),
+                                     phaseSpec.isStageLoggingEnabled(), phaseSpec.getConnectorDatasets()));
+
+    JavaPairRDD<String, Object> sourceTransformed =
+      input.flatMapToPair(new MapFunction<>(context, sourcePipelineStr, null)).cache();
+
+    SparkTransform sparkTransform =
+      new PipelinePluginInstantiator(context.getPluginContext(), phaseSpec).newPluginInstance(sparkTransformName);
+    JavaRDD<Object> sparkTransformed =
+      sparkTransform.transform(new BasicSparkPluginContext(context, null, sparkTransformName),
+                               sourceTransformed.values());
+
+    // phase starting from spark transform to sink(s)
+    PipelinePhase sinkPhase = phaseSpec.getPhase().subsetFrom(ImmutableSet.of(sparkTransformName));
+    String sinkPipelineStr =
+      GSON.toJson(new BatchPhaseSpec(phaseSpec.getPhaseName(), sinkPhase, phaseSpec.getResources(),
+                                     phaseSpec.isStageLoggingEnabled(), phaseSpec.getConnectorDatasets()));
+
+    JavaPairRDD<String, Object> sinkTransformedValues =
+      sparkTransformed.flatMapToPair(new SingleTypeRDDMapFunction(context, sinkPipelineStr)).cache();
+    return sinkTransformedValues;
+  }
 
   /**
    * Base function that knows how to set up a transform executor and run it.
    * Subclasses are responsible for massaging the output of the transform executor into the expected output,
    * and for configuring the transform executor with the right part of the pipeline.
    */
-  public abstract static class TransformExecutorFunction<KEY_IN, VAL_IN, KEY_OUT, VAL_OUT>
-    implements PairFlatMapFunction<Tuple2<KEY_IN, VAL_IN>, KEY_OUT, VAL_OUT> {
+  public abstract static class TransformExecutorFunction<IN, EXECUTOR_IN, KEY_OUT, VAL_OUT>
+    implements PairFlatMapFunction<IN, KEY_OUT, VAL_OUT> {
 
     protected final PluginContext pluginContext;
     protected final Metrics metrics;
     protected final long logicalStartTime;
     protected final Map<String, String> runtimeArgs;
     protected final String pipelineStr;
-    private transient TransformExecutor<KeyValue<KEY_IN, VAL_IN>> transformExecutor;
+    private transient TransformExecutor<EXECUTOR_IN> transformExecutor;
 
-    public TransformExecutorFunction(SparkContext sparkContext) {
+    public TransformExecutorFunction(SparkContext sparkContext, @Nullable String pipelineStr) {
```

Comment:
I think its simpler to indicate that null means to get it directly from the context.
We only pass it if we want to override it.